src/phylo/phylogeny.jl

```diff
@@ -0,0 +1,230 @@
+
+Construct a completely unresolved phylogeny, with no annotated data.
+
+The phylogeny will store annotation data of type C for tree nodes
+(clades and tips). It will also store data of type B associated with branches.
+
+# Examples
+```julia
+Phylogeny([:Human, :Chimp, :Dog, :Cat], Void, Void)
+Phylogeny([:Human, :Chimp, :Dog, :Cat], Float64, Float64)
+```
+"""
+function Phylogeny{C, B}(taxa::Vector{Symbol}, ::Type{C}, ::Type{B})
+    ntaxa = length(taxa)
+    maxVertices = max_vertices(ntaxa)
+    taxaAndRoot = push!(copy(taxa), :Root)
+    idxer = Indexer(taxaAndRoot, Int)
+    g = DiGraph(maxVertices)
+    for i in 1:ntaxa
+        add_edge!(g, ntaxa + 1, i)
+    end
+    v = Vector{C}()
+    d = Dict{Edge, B}()
+    return Phylogeny{C, B}(g, idxer, v, d, ntaxa, false, true)
+end
+
+
+"""
+    Phylogeny(taxa::Vector{Symbol})
+
+Construct a completely unresolved phylogeny, with no annotated data.
+
+This constructor assumes you want a phylogeny where the only data associated
+with it is branch lengths (Float64) for the edges, and confidence values
+(Float64) for the clades.
+
+# Examples
+```julia
+Phylogeny([:Human, :Chimp, :Dog, :Cat])
+```
+"""
+function Phylogeny(taxa::Vector{Symbol})
+    return Phylogeny(taxa, Float64, Float64)
+end
+
+
+"""
+    Phylogeny{T <: AbstractString}(taxa::Vector{T})
+Construct a completely unresolved phylogeny, with no annotated data.
+
+This constructor assumes you want a phylogeny where the only data associated
+with it is branch lengths (Float64) for the edges, and confidence values
+(Float64) for the clades.
+"""
+function Phylogeny{T <: AbstractString}(taxa::Vector{T})
+    return Phylogeny(convert(Vector{Symbol}, taxa))
+end
+
+
+function Base.copy{C, B}(tree::Phylogeny{C, B})
+    g = copy(tree.graph)
+    i = copy(tree.vertexindex)
+    v = copy(tree.vertexdata)
+    d = copy(tree.edgedata)
+    return Phylogeny{C, B}(g, i, v, d, tree.ntaxa, tree.rooted, tree.rerootable)
+end
+
+
+# Miscellaneous Functions
+# ------------------------
+
+function n_possible_rooted(n::Int)
+    @assert n >= 2 "n must be greater or equal to 2."
+    num = factorial(2 * n - 3)
+    den = (2^(n - 2)) * factorial(n - 2)
+    return Int(num / den)
+end
+
+function n_possible_unrooted(n::Int)
+    @assert n >= 3 "n must be greater of equal to 3."
+    return n_possible_rooted(n - 1)
+end
+
+max_vertices(n::Int) = 2 * n - 1
+
+
+typealias SymString Union{Symbol, AbstractString}
+
+
+"""
+    is_rooted(x::Phylogeny)
+
+Test whether a Phylogeny is rooted.
+
+# Examples
+```julia
+is_rooted(my_phylogeny)
+```
+"""
+function is_rooted(x::Phylogeny)
+    return x.rooted
+end
+
+
+"""
+    is_rerootable(x::Phylogeny)
+
+Test whether a Phylogeny is re-rootable.
+
+# Examples
+```julia
+is_rerootable(my_phylogeny)
+```
+"""
+function is_rerootable(x::Phylogeny)
+    return x.rerootable
+end
+
+
+"Get the vertex of the tree which represents the root of the tree."
+root_vertex(tree::Phylogeny) = tree.ntaxa + 1
```

Comment:
It is intended as it is common in all phylogenetic software that the trees have a "root" even if the tree is considered an unrooted tree, typically for practical reasons. A rooted and unrooted tree will always have a vertex that is reserved for a root (ntaxa + 1). So this function can be seen as getting the vertex that is reserved for the root of the tree, regardless of whether or not the tree is rooted and the vertex is connected. Similar to how clade_vertices returns the vertices that the tree has to use as clade nodes, regardless of whether or not they are used or connected up to the tree structure.